contracts/Subscription.sol

```diff
@@ -1,205 +1,356 @@
+    /// @dev This function can only be called by the admin.
+    /// @param duration The maximum duration (in seconds) for subscription renewal.
+    function setMaxRenewalDuration(uint64 duration) external onlyAdmin {
+        _setMaximumRenewalDuration(duration);
     }
 
-    function isSubscriptionValid(uint256 tokenId) public view returns (bool) {
-        return _expirations[tokenId] > block.timestamp;
+    /*//////////////////////////////////////////////////////////////
+                   ONLY-MINTER NON-CONSTANT FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /*//////////   updateSubscriptionForFree variants   //////////*/
+
+    /// @inheritdoc ISubscription
+    function updateSubscriptionForFree(uint256 tokenId, uint64 duration) external override onlyMinter {
+        _validateDurationBetweenMinAndMax(duration);
+        _updateSubscriptionExpiration(tokenId, duration);
     }
 
-    function renewSubscription(uint256 tokenId, uint64 duration) external payable virtual override {
-        // console2.log("msg.sender: ", msg.sender);
-        // console2.log("tokenId: ", tokenId);
-        // console2.log("duration: ", duration);
+    /// @inheritdoc ISubscription
+    function updateSubscriptionForFree(uint256[] calldata tokenIds, uint64 duration) external override onlyMinter {
+        uint256 tokenId;
 
-        // only owner of the token id can call this
-        // isApprovedOrOwner in crea8ors ??
-        // or write a custom function ??
-        if (msg.sender != erc721.ownerOf(tokenId)) {
-            revert CallerNotOwnerNorApproved();
-        }
+        for (uint256 i = 0; i < tokenIds.length;) {
+            tokenId = tokenIds[i];
+
+            _validateDurationBetweenMinAndMax(duration);
+            _updateSubscriptionExpiration(tokenId, duration);
 
-        // check duration
-        if (duration < _minimumRenewalDuration) {
-            revert RenewalTooShort();
-        } else if (_maximumRenewalDuration != 0 && duration > _maximumRenewalDuration) {
-            revert RenewalTooLong();
+            unchecked {
+                ++i;
+            }
         }
+    }
 
-        // check msg.value
-        if (msg.value < _getRenewalPrice(tokenId, duration)) {
-            revert InsufficientPayment();
+    /// @inheritdoc ISubscription
+    function updateSubscriptionForFree(
+        uint256[] calldata tokenIds,
+        uint64[] calldata durations
+    )
+        external
+        override
+        onlyMinter
+    {
+        if (tokenIds.length != durations.length) revert LengthMismatch();
+
+        uint256 tokenId;
+        uint64 duration;
+
+        for (uint256 i = 0; i < tokenIds.length;) {
+            tokenId = tokenIds[i];
+            duration = durations[i];
+
+            _validateDurationBetweenMinAndMax(duration);
+            _updateSubscriptionExpiration(tokenId, duration);
+
+            unchecked {
+                ++i;
+            }
         }
+    }
```

Comment:
we shouldn't need this function for array of `duration[]`, duration should be constant, even if an array of tokenIds is passed.

`duration` will always be `365 days` for mint day context on 8/8 mint day